README.md
# bilibili-live-effective-rate

**Due to Bilibili, this project is dead.**

---

**effective-watching-time** / **live-duration** = **live-effective-rate**

[简体中文 README](README_zh-cmn-Hans.md)

Greasy Fork: https://greasyfork.org/zh-CN/scripts/421252-bilibili-live-effective-rate

![bilibili-live-effective-rate screenshot_sessionData](screenshot_sessionData.png "bilibili-live-effective-rate screenshot_sessionData")

This project is licensed under **GNU AFFERO GENERAL PUBLIC LICENSE Version 3**


CHANGELOG.md
# bilibili-live-effective-rate Changelog

**Due to Bilibili, this project is dead.**

---

(From new to old)

## 0.4.0

The bilibili live data removed the row '最高人气值'.

## 0.3.0

The bilibili live data is changed from <https://link.bilibili.com/p/center/index#/data/overview> to <https://link.bilibili.com/p/center/index#/live-data/session-data>. A lot of things have changed. The v0.2.x branch works no more.

Notice the 0.3.0 version is still using **MutationObserver** which is not necessary any more, but I'm lazy to get rid of them.

The detail tab is removed from bilibili so will no longer work there.

Otherwise, there isn't much changed to talk about.

------

## 0.2.1

- Add unit 'second' to avoid some problems.

## 0.2.0

I think what I can do is done for now. There would be no more feature update.

- Removed unnecessary meta(greasyfork.org request)
- Update Url Checking
- Change color pattern

### About these:

#### Well for the second one

Sometimes, the url becomes to `https://link.bilibili.com/p/center/index?visit_id=40utwbkde8u8#/data/overview`. I have no idea when or why there is a get_param there, but it would break the old way of url matching checking.

#### The third one

Streamer watching is also counted as effective-watching-time, so the color pattern is changed to this:

- when the effectiveRate < 50, color = 'darkgreen'.
- when the 50 <= effectiveRate < 100, color = 'green'. (It means you have not a single person watched your live-streaming, including yourself)
- when the 100 <= effectiveRate < 150, color = 'darkgoldenrod'.
- when the 150 <= effectiveRate < 200, color = 'orange'. (It means you have not a single person watched your live-streaming, excluding yourself)
- when the effectiveRate >= 200, color = 'red'. (It means you have more than a single person watched your live-streaming, excluding yourself. Congratulation.)

## 0.1.0

- **Add popup-panel support**
- Make unitConverter functional
- Make colorRate functional
- Clean code

## 0.0.1

First release.


bilibili-live-effective-rate.js
// ==UserScript==
// @name                bilibili-live-effective-rate
// @name:zh-HK          bilibili-live-effective-rate
// @name:zh-TW          bilibili-live-effective-rate
// @name:en             bilibili-live-effective-rate
// @name:ja             bilibili-live-effective-rate
// @namespace           https://github.com/catscarlet/bilibili-live-effective-rate
// @description         bilibili-live-effective-rate 有效观看时长÷直播时长=直播有效率
// @description:zh-HK   bilibili-live-effective-rate 有效觀看時長÷直播時長=直播有效率
// @description:zh-TW   bilibili-live-effective-rate 有效觀看時長÷直播時長=直播有效率
// @description:en      bilibili-live-effective-rate Effective-watching-time / live-duration = live-effective-rate
// @description:ja      bilibili-live-effective-rate 有効観看時長÷直播時長=直播有効率
// @version             0.4.0
// @author              catscarlet
// @match               *://link.bilibili.com/p/center/index*
// @require             https://code.jquery.com/jquery-latest.js
// @run-at              document-end
// @grant               none
// @license             GNU AGPLv3
// ==/UserScript==

/*
This project is licensed under **GNU AFFERO GENERAL PUBLIC LICENSE Version 3**
*/

(function() {
    'use strict';

    $(function() {

        console.log('bilibili-live-effective-rate loaded');

        checkUrl(document.location);

        window.onpopstate = function(event) {
            checkUrl(document.location);
        };

        function checkUrl(url) {
            let urlRe = new RegExp('https://link\.bilibili\.com/p/center/index\\??\\S*#/live-data/session-data');
            let result = urlRe.test(url);
            if (result) {
                findCenterGrid();
                //findLinkPopupCtnr();
            }
        }

        function findCenterGrid() {
            const targetNode = document.getElementsByClassName('main-ctnr p-relative f-right')[0];
            const config = {attributes: false, childList: true, subtree: true};
            const callback = function(mutationsList, observer) {
                for (let mutation of mutationsList) {
                    if (mutation.type === 'childList') {
                        addAction();
                    } else if (mutation.type === 'attributes') {
                    }
                }
            };
            const observer = new MutationObserver(callback);
            observer.observe(targetNode, config);
        };

        /*
        function findLinkPopupCtnr() {
            const targetNode = document.getElementsByClassName('link-popup-ctnr')[0];
            const config = {attributes: false, childList: true, subtree: true};
            const callback = function(mutationsList, observer) {
                for (let mutation of mutationsList) {
                    if (mutation.type === 'childList') {
                        addActionToPopup();
                    } else if (mutation.type === 'attributes') {
                    }
                }
            };
            const observer = new MutationObserver(callback);
            observer.observe(targetNode, config);
        };
        */

        function colorRate(effectiveRate) {
            let color;
            if (effectiveRate < 50) {
                color = 'darkgreen';
            } else if (effectiveRate < 100) {
                color = 'green';
            } else if (effectiveRate < 150) {
                color = 'darkgoldenrod';
            } else if (effectiveRate < 200) {
                color = 'orange';
            } else {
                color = 'red';
            }

            return color;
        };

        function unitConverter(raw) {
            let duration = raw.slice(-2);
            let durationInMinute = 0;

            if (duration == '小时') {
                durationInMinute = parseFloat(raw.substr(0, raw.length - 2)) * 60;
            } else if (duration == ' 秒') {
                durationInMinute = parseFloat(raw.substr(0, raw.length - 2)) / 60;
            } else {
                durationInMinute = parseFloat(raw.substr(0, raw.length - 2));
            }

            return durationInMinute;
        };

        function addAction() {
            let sessionDataTable = $('.session-data-table');


            if (sessionDataTable.length == 0) {
                return;
            }

            if (sessionDataTable.attr('effectiveRateFlag')) {
                return;
            }

            let effectiveHead = '<td class="effective-rate-head" style="width: 10em">有效率</td>';

            let sessionDataTableThead = sessionDataTable.find('thead');

            let sessionDataTableTheadTrList = sessionDataTableThead.find('tr');
            sessionDataTableTheadTrList.each(function(i) {
                let tr = $(this);
                let tdList = tr.find('td');

                let sessionDataTableTheadAverageTimeDom = $(tdList[6]);

                sessionDataTableTheadAverageTimeDom.after(effectiveHead);

            });


            let sessionDataTableTbody = sessionDataTable.find('tbody');
            let sessionDataTableTbodyTrList = sessionDataTableTbody.find('tr');

            sessionDataTableTbodyTrList.each(function(i) {
                let tr = $(this);
                let tdList = tr.find('td');

                let durationDom = $(tdList[2]);
                let averageTimeDom = $(tdList[6]);

                let durationRaw = durationDom.text();
                let averageTimeRaw = averageTimeDom.text();

                let duration = unitConverter(durationRaw);
                let averageTime = unitConverter(averageTimeRaw);

                let effectiveRate = (averageTime / duration * 100).toFixed(2);
                let effectiveRatePercent = effectiveRate + ' %';

                let color = colorRate(effectiveRate);

                let effectiveRateTd = '<td data-v-045de783="" class="effective-rate" style="width: 10em; color: ' + color + '">' + effectiveRatePercent + '</td>';

                averageTimeDom.after(effectiveRateTd);
            });

            sessionDataTable.attr('effectiveRateFlag', 1);
        }

        /*
        function addActionToPopup() {
            let linkPopupPanel = $('.link-popup-panel');
            linkPopupPanel.css('width', 'max-content');

            if (linkPopupPanel.length == 0) {
                console.log('no linkPopupPanel. Return');
                return;
            }

            if (linkPopupPanel.attr('effectiveRateFlag')) {
                console.log('linkPopupPanel effectiveRateFlag. Return');
                return;
            }

            let linkPopupPanelEffectiveHead = '<span class="effective-rate-head" style="display: inline-block; line-height: 38px; font-size: 12px; text-align: center;">有效率</td>';

            let linkPopupPanelTitle = linkPopupPanel.find('.title');
            linkPopupPanelTitle.find('.average-time').after(linkPopupPanelEffectiveHead);

            let linkPopupPanelContent = linkPopupPanel.find('.content');
            let linkPopupPanelContentList = linkPopupPanelContent.find('.item');
            linkPopupPanelContentList.each(function(i) {
                let linkPopupPanelContentItem = $(this);

                let durationRaw = linkPopupPanelContentItem.find('.duration').text();
                let averageTimeRaw = linkPopupPanelContentItem.find('.average-time').text();

                let duration = unitConverter(durationRaw);
                let averageTime = unitConverter(averageTimeRaw);

                let effectiveRate = (averageTime / duration * 100).toFixed(2);
                let effectiveRatePercent = effectiveRate + ' %';

                let color = colorRate(effectiveRate);

                let effectiveRateSpan = '<span class="effective-rate-item" style="color: ' + color + '; display: inline-block; line-height: 38px; font-size: 12px; text-align: center;">' + effectiveRatePercent + '</td>';
                linkPopupPanelContentItem.find('.average-time').after(effectiveRateSpan);
            });

            linkPopupPanel.attr('effectiveRateFlag', 1);
        }
        */

    });
})();

/*
This project is licensed under **GNU AFFERO GENERAL PUBLIC LICENSE Version 3**
*/
